### PR TITLE
Ip elf

### PR DESCRIPTION
**Pull request by Heatdh**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -160,7 +160,7 @@ ELSE()
     SET(ETISS_USE_DLSYM ON)
     SET(ETISS_USE_DLSYM_DEEPBIND ON)
     SET(ETISS_USE_GETPROC OFF)
-    SET(ETISS_USE_BYTESWAP_H OFF)
+    SET(ETISS_USE_BYTESWAP_H ON)
```
Comment (rafzi) [inline CMakeLists.txt:163]: should be reverted

File: CMakeLists.txt [hunk 2]
```diff
@@ -173,7 +173,7 @@ ELSE()
         ETISS_CAN_USE_DLSYM_DEEPBIND
     )
     IF(NOT ETISS_CAN_USE_DLSYM_DEEPBIND)
-        #MESSAGE(FATAL_ERROR "Need RTLD_DEEPBIND on Linux")
+        MESSAGE(FATAL_ERROR "Need RTLD_DEEPBIND on Linux")
```
Comment (rafzi) [inline CMakeLists.txt:176]: should be reverted

File: CMakeLists.txt [hunk 3]
```diff
@@ -185,7 +185,7 @@ ELSE()
     ENDIF()
     CHECK_INCLUDE_FILES(byteswap.h ETISS_CAN_USE_BYTESWAP_H)
     IF(NOT ETISS_CAN_USE_BYTESWAP_H)
-        #MESSAGE(FATAL_ERROR "Need byteswap on Linux")
+        MESSAGE(FATAL_ERROR "Need byteswap on Linux")
```
Comment (rafzi) [inline CMakeLists.txt:188]: should be reverted

File: examples/bare_etiss_processor/main.cpp [hunk 4]
```diff
@@ -83,12 +84,17 @@ int main(int argc, const char *argv[])
     //        ETISS_SRCLOC);
     //  }
 
-    if (!dsys.loadRom(etiss::cfg().get<std::string>("sw_binary_rom", "").c_str()) ||
-        !dsys.loadRam(etiss::cfg().get<std::string>("sw_binary_ram", "").c_str()))
-    {
-        etiss::log(etiss::FATALERROR, "ram/rom not loaded properly\n");
+//     if (!dsys.loadRom(etiss::cfg().get<std::string>("sw_binary_rom", "").c_str()) ||
```
Comment (rafzi) [inline examples/bare_etiss_processor/main.cpp:87]: should be removed instead of commented out. no code is lost since we use version control

File: examples/bare_etiss_processor/main.cpp [hunk 5]
```diff
@@ -83,12 +84,17 @@ int main(int argc, const char *argv[])
...
+//     {
+//         etiss::log(etiss::FATALERROR, "ram/rom not loaded properly\n");
+//     }
+    
+    if (dsys.load_elf(etiss::cfg().get<std::string>("vp::elf_file", "").c_str() ) < 0 ){
+      etiss::log(etiss::FATALERROR, "ELF file not loaded properly\n");
     }
-
+    //uint8_t tbuf[4] = {};
+    //dsys.add_memsegment(make_unique<etiss::MemSegment>(etiss::cfg().get<int>("logaddr", 0x80000000), 4, etiss::MemSegment::WRITE, std::string("logger") ), tbuf);
```
Comment (rafzi) [inline examples/bare_etiss_processor/main.cpp:97]: remove testing code

File: examples/bare_etiss_processor/main.cpp [hunk 6]
```diff
@@ -113,22 +119,27 @@ int main(int argc, const char *argv[])
...
-        break;
-    }
 
+		etiss::uint64 startAddress = dsys.get_startaddr();
+		std::cout << "ELF start address: 0x" << std::hex << startAddress << std::dec << std::endl;
+
+//     etiss::uint64 startAddress = 0x0;
+//      switch ((char)CPUArchName.c_str()[0])
+//      {
+//      case (char)'o':
```
Comment (rafzi) [inline examples/bare_etiss_processor/main.cpp:129]: remove

File: examples/bare_etiss_processor/main.cpp [hunk 7]
```diff
@@ -113,22 +119,27 @@ int main(int argc, const char *argv[])
...
-        break;
-    // TODO: For RISCV, MSTATUS register has to be configured in advance to
-    // support interrupt.
-    case (char)'R':
-        startAddress = 0x80;
-        break;
-    }
 
+		etiss::uint64 startAddress = dsys.get_startaddr();
+		std::cout << "ELF start address: 0x" << std::hex << startAddress << std::dec << std::endl;
```
Comment (rafzi) [inline examples/bare_etiss_processor/main.cpp:124]: fix indentation; tabs -> spaces

File: examples/bare_etiss_processor/run_helper.sh.in [hunk 8]
```diff
@@ -112,4 +103,3 @@ else
     ${VP_EXE} ${ARGS} ${CMD_OPTIONS}
 fi
 
-rm $DYN_INI
```
Comment (rafzi) [inline examples/bare_etiss_processor/run_helper.sh.in:115]: add this line back

File: include/etiss/DebugSystem.h [hunk 9]
```diff
@@ -53,16 +53,90 @@
...
+      if(mem){ // use reserved memory
+        mem_ = mem;
+      } else {
+        mem_ = new etiss::uint8[size];
+        self_allocated_ = true;  
+      }
+    }
+    
+    virtual ~MemSegment(void){
+        if (self_allocated_ == true )
```
Comment (rafzi) [inline include/etiss/DebugSystem.h:103]: simplify to `if (self_allocated_)`

File: include/etiss/DebugSystem.h [hunk 10]
```diff
@@ -53,16 +53,90 @@
...
+    }
+    
+    void load(const void* data, size_t file_size_bytes){
+      if(data != nullptr and file_size_bytes <= size_){
+        memcpy(mem_, data, file_size_bytes);
+      }
+    }
+    
+    inline bool addr_in_range(etiss::uint64 addr) const {
+      return ( (addr >= start_addr_ and addr <= end_addr_) ? true : false);
```
Comment (rafzi) [inline include/etiss/DebugSystem.h:114]: simplify to just the condition because it already evaluated to true or false

File: include/etiss/DebugSystem.h [hunk 11]
```diff
@@ -53,16 +53,90 @@
...
+    }
+    
+    inline bool payload_in_range(etiss::uint64 addr, etiss::uint64 payload_size) const {
+      if(addr_in_range(addr)){
+        return ( ((addr+payload_size - 1) <= end_addr_) ? true : false);
```
Comment (rafzi) [inline include/etiss/DebugSystem.h:119]: simplify to just the condition because it already evaluated to true or false

File: include/etiss/DebugSystem.h [hunk 12]
```diff
@@ -53,16 +53,90 @@
...
         @brief simple etiss:System implementation for testing
 */
 class DebugSystem : public System
 {
   public:
+
     DebugSystem(uint32_t rom_start, uint32_t rom_size, uint32_t ram_start, uint32_t ram_size);
+    DebugSystem(void);
+
+    virtual ~DebugSystem(void){
```
Comment (rafzi) [inline include/etiss/DebugSystem.h:137]: not required; remove

File: src/DebugSystem.cpp [hunk 13]
```diff
@@ -69,12 +73,107 @@ uint32_t printMessage(std::string key, std::string message, uint32_t maxCount)
...
+      etiss::cfg().set<std::string>("CPUArch", "RISCV64");// RISCV and OR1K work as well 
+    if ((reader.get_class() == ELFCLASS32))
+      etiss::cfg().set<std::string>("CPUArch", "RISCV");
+      //add conditions 
+    else 
+      std::cout<<"System architecture is neither 64 nor 32 bit"<<std::endl;
+
+  }
+  // 
+  if (reader.get_machine() == EM_OPENRISC) {
```
Comment (rafzi) [inline src/DebugSystem.cpp:95]: join the conditionals with an else if ...

File: src/DebugSystem.cpp [hunk 14]
```diff
@@ -69,12 +73,107 @@ uint32_t printMessage(std::string key, std::string message, uint32_t maxCount)
...
+    if ((reader.get_class() == ELFCLASS32))
+        etiss::cfg().set<std::string>("CPUArch", "OR1K");
+    if ((reader.get_class() == ELFCLASS64))
+        std::cout<<"OR1k 64 is not supported ";
+  }
+  if ((reader.get_machine() != EM_OPENRISC) && (reader.get_machine() != EM_RISCV))
```
Comment (rafzi) [inline src/DebugSystem.cpp:101]: ... so that this can just be an else, to avoid the duplicate check

File: src/DebugSystem.cpp [hunk 15]
```diff
@@ -69,12 +73,107 @@ uint32_t printMessage(std::string key, std::string message, uint32_t maxCount)
...
+  , ram_start_(ram_start)
+  , rom_size_(rom_size)
+  , ram_size_(ram_size)
 {
     // mem = new etiss::uint8[DEBUGSYSTEM_MEMBLOCKSIZE*2];
-    rom_mem.resize(rom_size, 0);
-    ram_mem.resize(ram_size, 0);
+    rom_mem_.resize(rom_size, 0);
+    ram_mem_.resize(ram_size, 0);
+    std::cout<<"Hey I m the wished constructor"<<std::endl;
```
Comment (rafzi) [inline src/DebugSystem.cpp:176]: remove debug output

File: src/DebugSystem.cpp [hunk 16]
```diff
@@ -88,369 +187,247 @@ DebugSystem::DebugSystem(uint32_t rom_start, uint32_t rom_size, uint32_t ram_sta
     }
 }
 
+
+DebugSystem::DebugSystem(void) :
```
Comment (rafzi) [inline src/DebugSystem.cpp:191]: avoid duplicate code by using delegating constructor:

```
DebugSystem::DebugSystem() : DebugSystem(-1, 0, -1, 0) {}
```

File: src/DebugSystem.cpp [hunk 17]
```diff
@@ -88,369 +187,247 @@ DebugSystem::DebugSystem(uint32_t rom_start, uint32_t rom_size, uint32_t ram_sta
...
 etiss::int32 DebugSystem::iread(ETISS_CPU *, etiss::uint64 addr, etiss::uint32 len)
 {
-    if (addr >= _rom_start && addr < _rom_start + rom_mem.size())
-    {
-        return RETURNCODE::NOERROR;
-    }
+  int i_seg = 0;
+  int n_segs = msegs_.size();
+  for(i_seg = 0; n_segs; ++i_seg){
+    if(msegs_[i_seg]->addr_in_range(addr)) break;
```
Comment (rafzi) [inline src/DebugSystem.cpp:216]: for large number of sections msegs_ should probably be a map, so we can search more efficiently, but the current code should actually be more efficient for the typical two segment binaries we have right now

File: src/ETISS.cpp [hunk 18]
```diff
@@ -426,10 +428,20 @@ void etiss::Initializer::loadIni(std::list<std::string> *files)
         std::cout << "Info: simpleIni already exists!" << std::endl;
 
     // load file
+    bool load_bol;
```
Comment (rafzi) [inline src/ETISS.cpp:431]: debug code not required anymore?

File: include/etiss/DebugSystem.h [hunk 19]
```diff
@@ -76,21 +150,35 @@ class DebugSystem : public System
             @brief loads a binary image from a file to the given address
     */
     // bool load(etiss::uint64 addr,const char * file);
-    bool loadRom(const char *file);
```
Comment (rafzi) [inline include/etiss/DebugSystem.h:79]: why were these removed? we still provide the interface and access functions for the ram/rom sections, but without these we cannot initialize the data

**Review by rafzi — CHANGES_REQUESTED** [review]

File: include/etiss/DebugSystem.h [hunk 20]
```diff
@@ -53,16 +53,90 @@
 #include "etiss/System.h"
 #include <fstream>
 
-namespace etiss
+#include <cstring>
+#include <iostream>
+#include <memory>
+
+template<typename T, typename... Args>
+std::unique_ptr<T> make_unique(Args&&... args)
```
Comment (rafzi) [inline include/etiss/DebugSystem.h:61]: use this instead: https://github.com/tum-ei-eda/etiss/blob/master/include/etiss/make_unique.h

File: src/DebugSystem.cpp (hunk 16 above)
Comment (rafzi) [reply src/DebugSystem.cpp]: FYI: what i meant is to completely remove the content of the function body, because it would do exactly the same as the other constructor. this code here executes the other constructor.

see: https://github.com/tum-ei-eda/etiss/pull/37/commits/c43174f23cb3848fe36c29c04130fccb098ebec9

**Comment by Heatdh** [comment]
@rafzi  Thanks a lot for the applied changes. Should I merge the pull request or leave it for the moment open  ?
